nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,177 @@
+                name=cmd.machine_name,
+                log_serial=logger.log_serial,
+                log_machinestate=logger.log_machinestate,
+                tmp_dir=tmp_dir,
+            ) for cmd in cmd(vm_scripts)
+        ]
+
+        @atexit.register
+        def clean_up() -> None:
+            with self.log.nested("cleaning up"):
+                for machine in self.machines:
+                    if not machine.release():
+                        continue
+                    self.log(f"killed {machine.name} (pid {machine.pid})")
+                for vlan in self.vlans:
+                    vlan.release()
+                    self.log(f"killed {vlan.nr} (pid {vlan.pid})")
+            self.log.release()
+
+    @contextmanager
+    def subtest(self, name: str) -> Iterator[None]:
+        """Group logs under a given test name
+        """
+        with self.log.nested(name):
+            try:
+                yield
+                return True
+            except:
+                self.log(f'Test "{name}" failed with error:')
+                raise
+
+    def export_symbols(self):
+        global machines
+        machines = self.machines
+        machine_eval = [
+            "global {0}; {0} = machines[{1}]".format(m.name, idx) for idx, m in enumerate(machines)
+        ]
+
+        exec("\n".join(machine_eval))
+
+        global start_all, test_script
+        start_all = self.start_all
+        test_script = self.test_script
+
+        global subtest
+
+        def subtest(name):
+            self.subtest(name)
+
+    def test_script(self) -> None:
+        """Run the test script from the environment ('testScript')
+        """
+        with self.log.nested("running the VM test script"):
+            exec(os.environ["testScript"])
+
+    def run_tests(self) -> None:
+        """Run all tests from the environment ('test')
+        or drop into a python repl for interactive execution
+        """
+        tests = os.environ.get("tests")
+        if tests is not None:
+            with self.log.nested("running the VM test script"):
+                exec(tests, globals())
+        else:
+            ptpython.repl.embed(
+                locals(), globals(),
+                configure=self.configure_python_repl)
+
+        # TODO: Collect coverage data
+
+        for machine in self.machines:
+            if machine.is_up():
+                machine.execute("sync")
+
+    def start_all(self) -> None:
+        """Start all machines
+        """
+        with self.log.nested("starting all VLans"):
+            for vlan in self.vlans:
+                vlan.start()
+        with self.log.nested("starting all VMs"):
+            for machine in self.machines:
+                machine.start()
+
+    def join_all(self) -> None:
+        """Wait for all machines to shut down
+        """
+        with self.log.nested("waiting for all VMs to finish"):
+            for machine in self.machines:
+                machine.wait_for_shutdown()
```

Comment:
by the way is there some pythonic "parallel map" that can do this in parallel?

Reply:
I guess not without subroutines / threads. I'll analyze if it could be worth it.

Reply:
We can bounce ideas back and forth in https://matrix.to/#/#nixpkgs-pr-125372:matrix.org . I don't know how to tackle this atm. Resolving meanwhile.